src/System.Management.Automation/engine/Utils.cs

```diff
@@ -33,6 +34,144 @@ namespace System.Management.Automation
     /// </summary>
     internal static class Utils
     {
+        private struct PrimitiveRange
+        {
+            internal readonly BigInteger MinValue;
+            internal readonly BigInteger MaxValue;
+
+            internal PrimitiveRange(BigInteger min, BigInteger max)
+            {
+                this.MinValue = min;
+                this.MaxValue = max;
+            }
+
+            internal void Deconstruct(out BigInteger min, out BigInteger max)
+            {
+                min = this.MinValue;
+                max = this.MaxValue;
+            }
+        }
+
+        private static readonly Dictionary<Type, PrimitiveRange> s_typeBounds = new Dictionary<Type, PrimitiveRange>()
+        {
+            { typeof(sbyte),   new PrimitiveRange(sbyte.MinValue,  sbyte.MaxValue)  },
+            { typeof(byte),    new PrimitiveRange(byte.MinValue,   byte.MaxValue)   },
+            { typeof(short),   new PrimitiveRange(short.MinValue,  short.MaxValue)  },
+            { typeof(ushort),  new PrimitiveRange(ushort.MinValue, ushort.MaxValue) },
+            { typeof(int),     new PrimitiveRange(int.MinValue,    int.MaxValue)    },
+            { typeof(uint),    new PrimitiveRange(uint.MinValue,   uint.MaxValue)   },
+            { typeof(long),    new PrimitiveRange(long.MinValue,   long.MaxValue)   },
+            { typeof(ulong),   new PrimitiveRange(ulong.MinValue,  ulong.MaxValue)  },
+            { typeof(decimal), new PrimitiveRange((BigInteger)decimal.MinValue, (BigInteger)decimal.MaxValue) },
+            { typeof(double),  new PrimitiveRange((BigInteger)double.MinValue,  (BigInteger)double.MaxValue)  },
+        };
+
+        internal static bool IsWithinTypeBounds(Type destinationType, double value)
+        {
+            if (!s_typeBounds.ContainsKey(destinationType))
+            {
+                return false;
+            }
+
+            (BigInteger minValue, BigInteger maxValue) = s_typeBounds[destinationType];
+            if (minValue > (BigInteger)Math.Round(value) || (BigInteger)Math.Round(value) > maxValue)
```

Comment:
Why can't we get the destination type from the external value? It seems to work... Though yes, I suppose I could expand them all out, it just seems like unnecessary code to do 10+ helper functions -- the casts would be required for only double or decimal I guess.

I should probably benchmark these methods and see what is quicker. I'm sure you're right, but I have to prove it for myself that the extra untidy code is worth it :D

Reply:
I think easiest way is to grab from CoreFX methods like Convert.ToInt32() and remove throw from them.

Reply:
Possibly? I'll have a looksee, but I think that might end up more complex for no real gain.

Reply:
@iSazonov Looking at the Convert.To(U)IntX() source, that would end up with significantly more helper methods (one for each type conversion) where otherwise we can simply have a method check bounds and return true/false. 

I will proceed with checking benchmarks of a battery of helper functions against the current method. 😄

Reply:
After re-review the code I discard my question about  Convert.To(U)IntX().